21795 - Subfolder with name "folder" gets deleted by test #testIssue6…

carlotxra opened the pull request:
…406 / FileDialog does not work with memory FileSystem

Fix https://pharo.fogbugz.com/f/cases/21795/Subfolder-with-name-folder-gets-deleted-by-test-testIssue6406-FileDialog-does-not-work-with-memory-FileSystem.

Fix FileDialogWindowTest to create and delete unique directory instead of a potential existing directory.
Add test and make changes to use memory FileSystem when using FileDialogWindow.
Many implicit references to Filesystem>>disk with methods like >>asFileReference etc.
Allow FileDialogWindow to handle root path.
Store the current instance of memory FileSystem so that a new one is not created on each call.

carlotxra committed:
21795 - Subfolder with name "folder" gets deleted by test #testIssue6406 / FileDialog does not work with memory FileSystem
Fix https://pharo.fogbugz.com/f/cases/21795/Subfolder-with-name-folder-gets-deleted-by-test-testIssue6406-FileDialog-does-not-work-with-memory-FileSystem.

Fix FileDialogWindowTest to create and delete unique directory instead of a potential existing directory.
Add test and make changes to use memory FileSystem when using FileDialogWindow.
Many implicit references to Filesystem>>disk with methods like >>asFileReference etc.
Allow FileDialogWindow to handle root path.

carlotxra committed:
Change calls to either use a new instance of memory FileSystem, or use current instance.
Delete all files and directories which were created from memory FileSystem

carlotxra commented:
Note that with these changes when we call FileSystem>>memory we answer with the current memory file system which in affect is a singleton (similar to how FileSystem>>disk works).
To create a new memory file system there is a method named FileSystem>>newMemory.

carlotxra commented:
This means that when using FileSystem>>memory we should treat it the same as if it was a disk io.e. clean up any files and directories in tests etc

jecisc commented:
I'm not sure it's a good idea to keep the same memory file system each time we call #memory. If we do that when do we garbage collect the files created in memory?

carlotxra committed:
Register MemoryStore to startup list so that we clear the current fs.

carlotxra commented:
Yes, you're right I was meant to register the MemoryStore to the startup list. We now will clear the in memory contents at startup.

jecisc commented:
This is a big change. I think we should maybe talk about it in the mailing list. Because for example, I work on seaside applications in which some files uploaded by the user are temporarily stored in the memory and in some deployment they are almost never restarted. If we are not aware of this we will have a memory leak.

jecisc commented:
Instead of changing #memory and adding a #newMemory, maybe we could keep the current #memory method as it is now and add a #currentMemoryFileSystem method keeping a singleton?

VincentBlondeau reviewed (changes requested):
I agree with @jecisc. This should be discussed on the mailing list.

guillep commented:
Hi, I skimmed the issue on the weekend but as soon as I saw the change of the semantics of `memory` and the addition of `newMemory` I said myself to wait for monday :).

What I'd suggest:
 - let's integrate everything except the modification of the `FileSystem>>memory` method. We can even integrate the new `FileSystem>>newMemory` without danger
 - then we can then discuss just THAT change in the mailing list

That would allow us to move forward, what do you think?

carlotxra committed:
Revert to creating new memory file system behaviour when calling FileSystem>>memory.
Have the ability to reference current memory file system when necessary.

carlotxra commented:
I've made the changes so that the previous behaviour of #memory is put back. Added a method #currentMemoryFileSystem if users want to reference the same current memory filesystem instance (useful for any logic where need to reference the same filesystem).
Removed #newMemory.

jecisc commented:
Thank you! I think it's much easier for current users. 

Still, I would do it in two pull requests to make it easier to understand for people that do not follow closely the Pharo integrations.

guillep commented on src/FileSystem-Memory/MemoryStore.class.st:
```diff
@@ -7,24 +7,49 @@ Class {
+{ #category : #'system startup' }
+MemoryStore class >> startUp [
+	self reset
+]
```
This is strange... Why having a singleton if we are going to reset it every time we save the image?

guillep reviewed (approved):
Change is ok, we can iterate it afterwards.

I do not yet "swallow" the singleton memory filesystem but so far it does no harm.

guillep commented:
Thanks!
